xarray/core/dataarray.py

```diff
@@ -2752,9 +2753,15 @@ def stack(
             If False, don't create any index.
             If None, create a multi-index only if exactly one single (1-d) coordinate
             index is found for every dimension to stack.
-        index_cls: class, optional
+        index_cls : class, optional
             Can be used to pass a custom multi-index type. Must be an Xarray index that
             implements `.stack()`. By default, a pandas multi-index wrapper is used.
+        invert : bool, default: False
+            When `True`, all dimensions of the DataArray except for the sequence of
+            dimensions specified in the `dimensions` parameter will be stacked.
+            `dimensions` must have a length of 1 in this case, all dimensions listed
+            must exist in the current DataArray. Neither the `**dimensions_kwargs` nor
+            the ellipsis syntaxes may be used.
```

Comment:
I empathize with the use case here. That said:

Low-ish confidence, but I'm not a great fan of a parameter which totally changes another parameter's meaning.

I'm not sure of a great way of doing it, though. Would adding this to `.groupby` possibly make sense?? That's generally how we "invert" the dimensions. Or another param `except_dimensions`?

Reply:
I agree that producing side-effects on other parameters is icky. I'm not sure how well it fits in `groupby` though... I'm open to an `except_dimensions` parameter... do you think it would need to be mutually exclusive with `dimensions`? Also open to something like a `stack_except` method, but don't know how sensitive the project is to introducing new methods.

Reply:
I think `except_dimensions` could be a reasonable option. I don't love it, since this would be the only place in the library we use it.

`.groupby` could maybe work:

original from the docstring:
```python
        >>> stacked = arr.stack(newdim=("x", "y"))
        >>> stacked
        <xarray.DataArray (z: 2, newdim: 6)>
        array([[ 0,  2,  4,  6,  8, 10],
               [ 1,  3,  5,  7,  9, 11]])
        Coordinates:
          * z        (z) <U5 'alpha' 'beta'
          * newdim   (newdim) object MultiIndex
          * x        (newdim) <U1 'a' 'a' 'a' 'b' 'b' 'b'
          * y        (newdim) int64 0 1 2 0 1 2
```

Using `groupby`:

```python
        >>> stacked = arr.groupby('z').stack('newdim')
        # or
        >>> stacked = arr.groupby('z').stack(newdim=...)
        >>> stacked
        <xarray.DataArray (z: 2, newdim: 6)>
        array([[ 0,  2,  4,  6,  8, 10],
               [ 1,  3,  5,  7,  9, 11]])
        Coordinates:
          * z        (z) <U5 'alpha' 'beta'
          * newdim   (newdim) object MultiIndex
          * x        (newdim) <U1 'a' 'a' 'a' 'b' 'b' 'b'
          * y        (newdim) int64 0 1 2 0 1 2
```

It does kinda make sense, since the `z` is the groupby key. And it's very consistent with how `groupby` generally works (i.e. try with `.sum` if this isn't clear). 

Depending on the construction, the arguments to `GroupBy.stack` are a bit different from those to the existing `.stack`.

On reflection, I quite like it. But it's quite possible I'm missing something.

Reply:
(deleted a comment I submitted too early)

I think `except_dimensions` could be a reasonable option. I don't love it, since this would be the only place in the library we use it.

`.groupby` could maybe work:

original from the docstring:
```python
        >>> stacked = arr.stack(newdim=("x", "y"))
        >>> stacked
        <xarray.DataArray (z: 2, newdim: 6)>
        array([[ 0,  2,  4,  6,  8, 10],
               [ 1,  3,  5,  7,  9, 11]])
        Coordinates:
          * z        (z) <U5 'alpha' 'beta'
          * newdim   (newdim) object MultiIndex
          * x        (newdim) <U1 'a' 'a' 'a' 'b' 'b' 'b'
          * y        (newdim) int64 0 1 2 0 1 2
```

Using `groupby`:

```python
        >>> stacked = arr.groupby('z').stack(newdim=...)
        >>> stacked
        <xarray.DataArray (z: 2, newdim: 6)>
        array([[ 0,  2,  4,  6,  8, 10],
               [ 1,  3,  5,  7,  9, 11]])
        Coordinates:
          * z        (z) <U5 'alpha' 'beta'
          * newdim   (newdim) object MultiIndex
          * x        (newdim) <U1 'a' 'a' 'a' 'b' 'b' 'b'
          * y        (newdim) int64 0 1 2 0 1 2
```

It does kinda make sense, since the `z` is the groupby key, and remains as a dimension like its general usage (i.e. try with `.sum` if this isn't clear).

On reflection, I quite like it. But it's quite possible I'm missing something.

Reply:
Does anyone else @pydata/xarray have views on the API? I think two broad options (along with a third of "do nothing, folks can write it themselves):
- The `groupby` option — potentially very nice, but also maybe too out there
- The `exclude_dims` option — this would be the only method we have this (the others use `.groupby`), so it's not really orthogonal

There's another which is to make `xr.ALL_DIMS` into a class, and allow `xr.ALL_DIMS.exclude('foo', 'bar')`. But introduces new functionality for possibly a fairly small use-case.

Reply:
Thanks @carschandler for the PR!

Maybe I'm too conservative, but do we really need some specific API for this case? Couldn't we just do

```python
arr.stack(newdim=[d for d in arr.dims if d != "z"])
```

A potential source of confusion for `Dataset.stack()` is the order of the stacked dimensions given as argument, which may be important in some cases especially if a multi-index is built from the stacked coordinates (it is by default). The order of the dimensions of a Dataset has no particular meaning so I think it is preferable to pass them in a more explicit way to `.stack`.

Reply:
Of course you could! I think you can always make this argument though (couldn't you just use `np.zeros + 1` instead of `np.ones`, `np.zeros(a.shape)` instead of `np.zeros_like`, and couldn't you just use a `MultiIndex` in a `DataFrame` to store multidimensional data?), and I’m in favor of adding conveniences when they don't obscure the rest of the code. If everyone agrees that this does indeed obscure the project, then I will gladly rest my case, but I think that the true value of a project like `xarray` is in how the data can be quickly and simply indexed, so I figured this would be a nice convenience since all the other ways of achieving it are kind of ugly. 

I do agree that the lack of order on a Dataset does pose a problem for setting the MultiIndex order... that being said, I think that there are still use cases where the order may not matter. Thanks for your feedback.